File: modules/09-computer-vision/ts/task4_convolution.ts
/**
 * task4_convolution.ts — 2-D convolution from scratch.  🔴
 *
 * What it teaches:
 *   Convolution is the fundamental operation in every CNN layer. A small matrix
 *   (the kernel or filter) slides over an image, computing a dot product at
 *   each position. The output — a feature map — highlights wherever the kernel's
 *   pattern matches the image.
 *
 *   Kernels you'll apply here:
 *     - Sobel horizontal: lights up horizontal edges (brightness changes top→bottom)
 *     - Sobel vertical:   lights up vertical edges   (brightness changes left→right)
 *     - Box blur:         averages each pixel with its 3×3 neighbourhood
 *
 *   In a trained CNN these kernels are *learned* from data, but they all work
 *   the same way: the conv inner loop you implement here is exactly what
 *   nn.Conv2d (PyTorch) or tf.keras.layers.Conv2D does under the hood (just
 *   massively parallelised on GPU).
 *
 *   We use raw TypedArrays (Float32Array) instead of a library so the
 *   computation is completely visible. No numpy, no tensor library.
 *
 * How to run (from the repo root):
 *   pnpm tsx modules/09-computer-vision/ts/task4_convolution.ts
 *
 *   Output: writes grayscale PGM files to assets/ so you can view them.
 *   (PGM is plain-text, viewable in any image viewer or by running:
 *    `open assets/edge_h.pgm` on macOS)
 *
 * Acceptance:
 *   - conv2d() passes the identity-kernel sanity check.
 *   - PGM files are written for each kernel.
 */

import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { mkdirSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import https from "node:https";
import fs from "node:fs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ASSETS_DIR = path.join(__dirname, "..", "assets");
const SAMPLE_IMAGE = path.join(ASSETS_DIR, "cat.jpg");

// ---------------------------------------------------------------------------
// Kernels (stored as flat row-major Float32Array + size)
// ---------------------------------------------------------------------------

interface Kernel {
  data: Float32Array;
  size: number; // kernels are size×size
}

function makeKernel(rows: number[][]): Kernel {
  const size = rows.length;
  const data = new Float32Array(size * size);
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      data[r * size + c] = rows[r][c];
    }
  }
  return { data, size };
}

const EDGE_HORIZONTAL = makeKernel([
  [-1, -2, -1],
  [ 0,  0,  0],
  [ 1,  2,  1],
]);

const EDGE_VERTICAL = makeKernel([
  [-1, 0, 1],
  [-2, 0, 2],
  [-1, 0, 1],
]);

const BLUR = makeKernel([
  [1/9, 1/9, 1/9],
  [1/9, 1/9, 1/9],
  [1/9, 1/9, 1/9],
]);

// ---------------------------------------------------------------------------
// Image representation — grayscale, row-major Float32Array
// ---------------------------------------------------------------------------

interface GrayImage {
  data: Float32Array;
  width: number;
  height: number;
}

// ---------------------------------------------------------------------------
// Task: implement 2-D convolution
// ---------------------------------------------------------------------------

/**
 * Apply a 2-D convolution kernel to a single-channel grayscale image.
 *
 * @param image  GrayImage with pixel values in [0, 255].
 * @param kernel Kernel to apply (must be odd-sized square).
 * @returns      GrayImage — the feature map, same spatial size as input.
 *
 * Algorithm:
 *   1. Allocate output Float32Array of size height × width.
 *   2. Compute kHalf = (kernel.size - 1) / 2 — the padding radius.
 *   3. For each output pixel (row, col):
 *      a. Accumulate the dot product of the kernel with the kernel.size×kernel.size
 *         patch centred at (row, col) in the input.
 *      b. For positions that fall outside the image boundary, treat them as 0
 *         (zero-padding). Check bounds before reading.
 *      c. Write the accumulated sum to output[row * width + col].
 *
 * Correlation vs convolution:
 *   We implement correlation (no kernel flip) — this is what torch.nn.Conv2d
 *   does too (the flip is absorbed into learned weights).
 */
function conv2d(image: GrayImage, kernel: Kernel): GrayImage {
  const { width, height } = image;
  const K = kernel.size;
  const kHalf = (K - 1) / 2;
  const output = new Float32Array(width * height);

  // TODO (exercise): implement the convolution loop (4 nested loops).
  //   - Outer two loops walk every output pixel (row, col).
  //   - Inner two loops walk the K×K kernel (kr, kc). Map each kernel cell to an
  //     input coordinate: `ir = row - kHalf + kr`, `ic = col - kHalf + kc` (this
  //     centres the kernel on the current pixel).
  //   - Guard the input read with a bounds check — if (ir, ic) falls outside the
  //     image, skip it, which treats it as 0 (zero-padding).
  //   - Accumulate `image.data[ir * width + ic] * kernel.data[kr * K + kc]` into a
  //     running `sum`, then store it at `output[row * width + col]`.
  //   - Remember both `image.data` and `output` are flat row-major arrays: index
  //     row r, col c as `r * width + c`.

  throw new Error(
    "Implement the convolution loop. " +
    "See the TODO hints above, then remove this throw."
  );

  return { data: output, width, height };
}

// ---------------------------------------------------------------------------
// Minimal JPEG decoder (via canvas-like approach using Node built-ins)
// We convert JPEG → PGM via a tiny manual parse — not production-grade.
// For a simpler approach: use `@huggingface/transformers` RawImage instead.
// ---------------------------------------------------------------------------

/**
 * Load a JPEG file and decode it to grayscale using transformers.js RawImage.
 * Falls back to a synthetic gradient image if loading fails.
 */
async function loadGrayscale(imagePath: string): Promise<GrayImage> {
  try {
    // transformers.js ships a RawImage utility that handles JPEG/PNG → RGB.
    const { RawImage } = await import("@huggingface/transformers") as any;
    const img = await RawImage.fromURL(`file://${imagePath}`);
    // Convert RGB to grayscale: Y = 0.299R + 0.587G + 0.114B
    const W = img.width as number;
    const H = img.height as number;
    const rgb = img.data as Uint8ClampedArray;
    const gray = new Float32Array(W * H);
    for (let i = 0; i < W * H; i++) {
      gray[i] = 0.299 * rgb[i * 3] + 0.587 * rgb[i * 3 + 1] + 0.114 * rgb[i * 3 + 2];
    }
    console.log(`Loaded   : ${imagePath} → ${W}×${H} grayscale`);
    return { data: gray, width: W, height: H };
  } catch {
    // Fallback: 64×64 synthetic gradient image.
    console.warn("Could not load JPEG — using a synthetic 64×64 gradient image.");
    const W = 64, H = 64;
    const data = new Float32Array(W * H);
    for (let r = 0; r < H; r++) {
      for (let c = 0; c < W; c++) {
        // Diagonal gradient with a hard edge at r === H/2
        data[r * W + c] = r < H / 2 ? (c / W) * 255 : 255 - (c / W) * 255;
      }
    }
    return { data, width: W, height: H };
  }
}

// ---------------------------------------------------------------------------
// Save as PGM (Portable GrayMap) — viewable in any image viewer
// ---------------------------------------------------------------------------

function saveAsPGM(image: GrayImage, filePath: string): void {
  const { data, width, height } = image;
  // Normalise to [0, 255]
  let max = 0;
  for (const v of data) if (Math.abs(v) > max) max = Math.abs(v);
  const scale = max > 0 ? 255 / max : 1;

  const pixels = new Uint8Array(width * height);
  for (let i = 0; i < pixels.length; i++) {
    pixels[i] = Math.min(255, Math.round(Math.abs(data[i]) * scale));
  }

  // PGM header + binary pixel data
  const header = `P5\n${width} ${height}\n255\n`;
  const buf = Buffer.concat([Buffer.from(header, "ascii"), Buffer.from(pixels)]);
  writeFileSync(filePath, buf);
  console.log(`Saved    : ${filePath}`);
}

// ---------------------------------------------------------------------------
// Sanity check
// ---------------------------------------------------------------------------

function sanityCheck(): void {
  // Identity kernel: convolution with it should return the image unchanged.
  const identity = makeKernel([[0, 0, 0], [0, 1, 0], [0, 0, 0]]);
  const img: GrayImage = {
    data: new Float32Array([1, 2, 3, 4, 5, 6, 7, 8, 9]),
    width: 3,
    height: 3,
  };
  const result = conv2d(img, identity);
  const ok = img.data.every((v, i) => Math.abs(v - result.data[i]) < 1e-5);
  if (!ok) {
    throw new Error(
      `Identity convolution failed!\nExpected: ${img.data}\nGot:      ${result.data}`
    );
  }
  console.log("Sanity check passed: identity convolution is correct.\n");
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function ensureSampleImage(): Promise<string> {
  mkdirSync(ASSETS_DIR, { recursive: true });
  if (existsSync(SAMPLE_IMAGE)) return SAMPLE_IMAGE;

  const url =
    "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Cat_November_2010-1a.jpg/320px-Cat_November_2010-1a.jpg";
  console.log(`Downloading sample image → ${SAMPLE_IMAGE}`);

  await new Promise<void>((resolve, reject) => {
    const file = fs.createWriteStream(SAMPLE_IMAGE);
    https.get(url, (res) => {
      res.pipe(file);
      file.on("finish", () => { file.close(); resolve(); });
    }).on("error", reject);
  });

  return SAMPLE_IMAGE;
}

async function main(): Promise<void> {
  sanityCheck();

  const imagePath = await ensureSampleImage();
  const gray = await loadGrayscale(imagePath);

  const kernels: Array<{ name: string; kernel: Kernel; desc: string }> = [
    { name: "edge_h", kernel: EDGE_HORIZONTAL, desc: "Sobel horizontal edges" },
    { name: "edge_v", kernel: EDGE_VERTICAL,   desc: "Sobel vertical edges" },
    { name: "blur",   kernel: BLUR,             desc: "3×3 box blur" },
  ];

  for (const { name, kernel, desc } of kernels) {
    console.log(`Applying : ${desc}`);
    const featureMap = conv2d(gray, kernel);
    saveAsPGM(featureMap, path.join(ASSETS_DIR, `${name}.pgm`));
  }

  console.log();
  console.log("Open the .pgm files in assets/ to visualise the feature maps.");
  console.log("Edge kernels light up boundaries; blur removes high-frequency detail.");
}

main().catch((err) => {
  console.error(err.message ?? err);
  process.exit(1);
});

File: modules/09-computer-vision/py/task4_convolution.py
"""task4_convolution.py — 2-D convolution from scratch.  🔴

What it teaches:
    Convolutional Neural Networks (CNNs) dominated computer vision from 2012
    (AlexNet) until Vision Transformers (ViT, 2020) took over. Understanding
    convolution is still essential because:
      - It's the operation inside every CNN layer.
      - ViT patches are a spatial structure too (just processed differently).
      - Edge detection by hand shows *exactly* what feature maps are.

    A 2-D convolution slides a small matrix (the **kernel** or **filter**) over
    an image, computing a dot product at each position. The output (a **feature
    map**) highlights whatever pattern the kernel encodes:
      - A horizontal-edge kernel lights up where brightness changes top→bottom.
      - A Gaussian-blur kernel smooths the image.
      - Learned CNN kernels detect curved edges, textures, eyes, wheels, etc.

    In a real CNN:
      - There are many kernels per layer (64, 256, …), each learning a different
        feature.
      - Multiple layers stack, so later kernels see combinations of earlier
        features (edges → shapes → parts → objects).

Kernels used here:
    EDGE_HORIZONTAL  — Sobel filter, detects horizontal edges
    EDGE_VERTICAL    — Sobel filter, detects vertical edges
    BLUR             — 3×3 box blur (averaging kernel)

How to run (from the repo root):
    uv sync --extra vision          # numpy + pillow (already present for task 1)
    uv run python modules/09-computer-vision/py/task4_convolution.py

    Output: saves edge_h.png / edge_v.png / blur.png next to the sample image.
    Open them to visualise what each kernel does.

Acceptance:
    - conv2d() passes the small sanity-check test at the bottom.
    - Output images are written and visually make sense (edges lit up, blur
      smoothed).

Note: numpy's own vectorised ops could do this faster; we use explicit Python
loops on purpose so the inner operation is completely transparent.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np

SAMPLE_IMAGE = Path(__file__).parent.parent / "assets" / "cat.jpg"
OUTPUT_DIR = Path(__file__).parent.parent / "assets"

# ---------------------------------------------------------------------------
# Kernels
# ---------------------------------------------------------------------------

# Sobel horizontal-edge kernel: bright where intensity changes top→bottom.
EDGE_HORIZONTAL = np.array([
    [-1, -2, -1],
    [ 0,  0,  0],
    [ 1,  2,  1],
], dtype=np.float32)

# Sobel vertical-edge kernel: bright where intensity changes left→right.
EDGE_VERTICAL = np.array([
    [-1, 0, 1],
    [-2, 0, 2],
    [-1, 0, 1],
], dtype=np.float32)

# Simple 3×3 box blur: each output pixel is the average of its 3×3 neighbourhood.
BLUR = np.ones((3, 3), dtype=np.float32) / 9.0


# ---------------------------------------------------------------------------
# Task: implement 2-D convolution
# ---------------------------------------------------------------------------

def conv2d(image: np.ndarray, kernel: np.ndarray) -> np.ndarray:
    """Apply a 2-D convolution kernel to a single-channel (grayscale) image.

    Args:
        image:  H × W float32 array, values in [0, 255].
        kernel: K × K float32 array (we only support odd-sized square kernels).

    Returns:
        H × W float32 array — the feature map. Edges use zero-padding so the
        output has the same spatial size as the input.

    Implementation notes:
        - k_half = kernel.shape[0] // 2 gives the padding on each side.
        - Pad `image` with zeros of width k_half on all four sides.
        - For each output position (i, j) extract the K×K patch from the padded
          image, element-wise multiply with the kernel, and sum the result.
        - That sum is output[i, j].

    Convolution vs correlation:
        Strictly, convolution flips the kernel 180° before sliding it; in deep
        learning the kernel is learned anyway so the flip is absorbed into the
        weights. Here we implement correlation (no flip), which is what PyTorch
        nn.Conv2d actually does.
    """
    H, W = image.shape
    K = kernel.shape[0]
    k_half = K // 2

    # Zero-pad the image so output is same size as input.
    padded = np.pad(image, k_half, mode="constant", constant_values=0)

    output = np.zeros((H, W), dtype=np.float32)

    # TODO (exercise): implement the convolution inner loop.
    #   - Loop over every output position (i, j) for i in range(H), j in range(W).
    #   - At each position, slice the K×K `patch` out of `padded` whose top-left
    #     corner is (i, j) — because `padded` is already offset by k_half, this
    #     patch is the neighbourhood centred on input pixel (i, j).
    #   - Element-wise multiply the patch by `kernel` and sum it (np.sum) into
    #     `output[i, j]`.

    raise NotImplementedError(
        "Implement the double for-loop that fills `output`. "
        "See the TODO comment above for guidance."
    )

    return output


# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------

def load_grayscale(path: Path) -> np.ndarray:
    """Load an image and convert to float32 grayscale [0, 255]."""
    try:
        from PIL import Image
    except ImportError:
        raise ImportError(
            "Pillow is not installed. Run: uv sync --extra vision\n"
            "or: pip install pillow"
        )
    img = Image.open(path).convert("L")  # "L" = 8-bit grayscale
    return np.array(img, dtype=np.float32)


def save_feature_map(feature_map: np.ndarray, path: Path) -> None:
    """Normalise a feature map to [0, 255] and save as a PNG."""
    from PIL import Image

    # Clip and normalise to [0, 255] so the image is viewable.
    fm = np.abs(feature_map)
    if fm.max() > 0:
        fm = fm / fm.max() * 255.0
    Image.fromarray(fm.astype(np.uint8)).save(path)
    print(f"Saved: {path}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    image_path = SAMPLE_IMAGE
    if not image_path.exists():
        image_path = _download_sample_image()

    print(f"Loading  : {image_path}")
    gray = load_grayscale(image_path)
    print(f"Shape    : {gray.shape}  (H × W, grayscale)")
    print()

    kernels = {
        "edge_h": (EDGE_HORIZONTAL, "Sobel horizontal edges"),
        "edge_v": (EDGE_VERTICAL,   "Sobel vertical edges"),
        "blur":   (BLUR,             "3×3 box blur"),
    }

    for name, (kernel, desc) in kernels.items():
        print(f"Applying : {desc}")
        feature_map = conv2d(gray, kernel)
        out_path = OUTPUT_DIR / f"{name}.png"
        save_feature_map(feature_map, out_path)

    print()
    print("Open the PNG files in assets/ to see the feature maps.")
    print("Notice: edge kernels light up boundaries; blur removes high-frequency detail.")


def _download_sample_image() -> Path:
    import urllib.request
    assets_dir = Path(__file__).parent.parent / "assets"
    assets_dir.mkdir(parents=True, exist_ok=True)
    dest = assets_dir / "cat.jpg"
    url = "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Cat_November_2010-1a.jpg/320px-Cat_November_2010-1a.jpg"
    print(f"Downloading sample image → {dest}")
    urllib.request.urlretrieve(url, dest)  # noqa: S310
    return dest


# ---------------------------------------------------------------------------
# Sanity check (run automatically when you execute this file)
# ---------------------------------------------------------------------------

def _sanity_check() -> None:
    """Quick test: convolving a 5×5 image with the identity kernel (centre=1,
    rest=0) should return the image unchanged."""
    identity = np.array([[0, 0, 0], [0, 1, 0], [0, 0, 0]], dtype=np.float32)
    img = np.arange(25, dtype=np.float32).reshape(5, 5)
    result = conv2d(img, identity)
    assert np.allclose(result, img), (
        f"Identity convolution failed!\nExpected:\n{img}\nGot:\n{result}"
    )
    print("Sanity check passed: identity convolution is correct.")


if __name__ == "__main__":
    _sanity_check()
    print()
    main()
